### ISSUE TITLE
Improvements on discovery map-view interface

### ISSUE THREAD

**Issue by xuezhaokun** [hydroshare/hydroshare#1361]
The succinct summary of these changes is as follows:
- [ ] - Don’t list the resource titles of coverages on the right. If possible, expand the map into that area. Instead, we will list them in a different way. See below.

- [ ] -  Don’t display centers of coverages. See below.

Then, implement a new mechanism for coverages as follows:
- [ ] -  Allow users to choose a specific rectangle to bound the search, independent of the zoom level. This could be a button “limit search to visible region” along with another button “remove search region limits” and “zoom to search llimit” that only appears when the search region is selected. These should be above the top of the map.

- [ ] - Make pan and zoom of the map independent of this function. If a search bound is selected, it appears as an unshaded rectangle (if possible, with a dotted edge) when users pan and zoom around it. Users can pan and zoom anywhere on the map and unless they “limit search to visible region” the search will be unchanged and unlimited.

- [ ] - List spatial boxes as icons above the map without titles, horizontally and possibly scrolling onto multiple lines. Each icon represents a specific coverage that is visible. These icons mean different things than before:

- Combine resources with exactly the same spatial coverage into one icon.

- Clicking on an icon pops up a (potentially scrollable) list of resources with that coverage. This should include “zoom to” and “show resource” clickable links for each resource. Only one of these lists pops up at a time.

- The icons shown are limited to those coverages that match search criteria and overlap with the current view area. There are no icons shown for regions with no search matches.

- The icons themselves are squares that are shaded to represent the shape of the coverage in the view area.

1. A rectangle with a black border and white center and dot for a small coverage.

2. A totally black rectangle if the coverage covers the whole view area.

3. Otherwise, a rectangle with a black border, white fill, and black embedded rectangle depicting the coverage and its relative position on the current view.  

- Sort these in order of decreasing area.

- Do not categorize by area – no area labels.

-  You might need a new javascript library to create these small images. See if you can create them clientside in the browser.  

 
Also, start thinking about how we might implement the following:
- [ ] - Allowing a user to draw an arbitrary polygon as a search bound.

- [ ] - Allowing import of the USGS watershed deliniations as search bounds.

**Comment by alvacouch** [comment]
There is another approach being discussed here. 

1.  Get rid of the right box and extend the map there.
2.  In the map view, replace the facets box with a coverage list with abbreviated resource titles in alphabetical order.

    a. eliminate articles: "a", "the", etc. 
    b. truncate the title to a few words. 
    c. alphabetize after that elimination.   

3. Split coverage resources into two types: 

   a. coverages that overlap but do not cover the current area. 
   b. coverages that completely cover the current area. 

4. Treat these two types differently:  Put them in different lists, and then:

   a. hovering over an overlapping coverage shows the coverage on the map. Clicking on it gives information on the resource. 
   b. hovering over a covering resource (whose coverage is more than the current area) does not show the coverage on the map. Clicking on it works similarly. See below for one mockup from @danames . 
@martinseul contributed to this. @dtarb: we continue to refine this idea. 

[figure.pdf](https://github.com/hydroshare/hydroshare/files/386145/figure.pdf)

Let's try this first before getting fancier......
A first step would be to modify the right-hand list to conform to this, and then move it to the left in a later step.

**Comment by xuezhaokun** [comment]
@alvacouch 
1. Do we have to show anything when we firstly initialize the map? 
2. If we initialize the map with markers clusters, when hover over a resource title, do we want to zoom in the cluster to show the corresponding symbol? Or do we just stay on current map view to show the symbol?
3. Do we automatically update the titles list as the map's bound changed?

**Comment by alvacouch** [comment]
Initialization stays the same. 

- Clusters indicate point resources. 
- Lists of coverages are all coverages matching the extent (globally) 
- we don't want to zoom when hover. We do want to be able to click on a coverage title and get a popup that allows us to zoom or visit the resource. 

We do automatically update the titles as the map changes.

**Comment by alvacouch** [comment]
@xuezhaokun This modification still assumes that we add buttons for limiting the search. Zooming does not change the search. It does change the coverages that are included, excluding the ones that don't overlap the current view region.

**Comment by xuezhaokun** [comment]
![image](https://cloud.githubusercontent.com/assets/6678333/17181042/a1ada02e-53ec-11e6-8405-28fc02906246.png)

For example, if hover over resource "random point" which is a point resource inside this cluster, do we want to show a marker above the cluster symbol?

**Comment by alvacouch** [comment]
@xuezhaokun We should do what we always did with point resources: 
a) no listing in the master directory above. 
b) a popup that describes the resources that is exposed when we get to a single resource. 
c) zooming to resolve clusters.

**Comment by xuezhaokun** [comment]
@alvacouch OK. That's clear. Since in the mockup figure, the instruction says we want to list all resources including point resources, it makes me feel some confused on how to deal with point resources.

**Comment by alvacouch** [comment]
@xuezhaokun I think it's a bit of overkill to do point resources in the list. We might do it eventually, but for now, let's do the point resources the way we always did. The open problem is how to deal with the coverages.

**Comment by xuezhaokun** [comment]
@alvacouch Got it!

**Comment by danames** [comment]
I'm experimenting with the current release on playground. Nice work implementing this, Zhaokun. This is very much what Alva and I discussed at the CUAHSI meeting. I do have a fairly large workflow modification request and a couple of smaller suggestions:

1) The workflow needs to be updated to be a bit more natural for the user. I've explained the updated workflow together with what it would look like in this figure: 
![image](https://cloud.githubusercontent.com/assets/3472628/17595629/fe71fa7a-5faa-11e6-8cd8-8207eb05459f.png)

2) Capture the "Enter" button click on the location search box and perform the zoom in search based on that click. Also change the name of the button "Zoom in" to "Go"
![image](https://cloud.githubusercontent.com/assets/3472628/17593780/412bfbce-5fa4-11e6-92b8-2b3d5afdeeb6.png)

3) Update the legend to include clusters and polygons, and also make the legend more compact (using less space): 
![image](https://cloud.githubusercontent.com/assets/3472628/17594375/5501e788-5fa6-11e6-9594-3d464147cf6c.png)

**Comment by xuezhaokun** [comment]
@danames Thanks for your suggestions, Dan! There are two things I am worried about this design. 
1. If we show all the boxes on the map, there might be a lot overlapping rectangles. It might make the map look messy.
2. For resources coverage the same area, we will have two exactly overlapping boxes in that area. When users click the box, it would be confused to decide which popup window should be shown. If we just show one box for such resources, and when users click it, we show a list of resources in this area, we would have the problem we had before. Users would have no idea about how many resources in this area, until they click the box to see the list.

Do you have any advice on these two issues?

**Comment by dtarb** [comment]
Here is an idea that may help.  If there are overlapping boxes (be they exactly overlapping or not) in the display and a user clicks in the overlap area, rather than show just the top one (as Dan suggested) pop up a box that shows up to 5 that overlap at the click point with a more option to open a window that shows all similar to the windows used for selecting resources for some of our apps (HydroShare GIS app specifically).  

I agree that with many resources the maps may get messy.  This can be an issue for point resources too.  Users will need to learn to filter on other facets as well as the map to resolve some of this.

**Comment by dtarb** [comment]
Zoom controls need adjusting.  If I go to playground and select discover and then my name in the Filter by owner facet, I get one red dot in the middle of a blue area.  It looks like the middle of the ocean.  However after zooming out a few levels I find I am in the middle of a small reservoir.  It should not have zoomed in so much.

Also when using drag zoom mode, keep it turned on until the user clicks on another tool or clicks to turn it off.  It is natural to zoom a bit, then want to drag zoom a bit more, and a nuisance to have to click the tool again.

Also, drag zoom mode does not appear to zoom to the area dragged.  It goes to an area somewhat bigger.

**Comment by xuezhaokun** [comment]
@dtarb 
- For the overlapping concern, for example, if we have 3 exactly overlapping resources, there would be only "one box" (three overlapping boxes) shown on the map. I think this might be confused for users. Users might see 3 resources from the list view, but on the map, they just see 1 box. They may think the display on map is wrong. The number of boxes on map does not match the number of results from list view. If they do not click the box to show the resource list, they will have no idea how many resources in this box. Do we have to take any care about this issue? Or we always want our users to click the box to find out what resources in the box. 
- For the zoom control problem, I made a function for calculating the zoom level based on the bound provided by markers. As we add a marker, we extend the map bound based on the marker's position. And finally based the map bound we get, we calculate a proper zoom level for it. For you case, since you just have one marker on the map, the bound generated by the marker is very small, and we will have a really high zoom level. That's why it zooms in so much.  
- As for drag zoom, I think it tries to center the drag area, then scale and zoom to a proper level. I don't think it is going to set the map bound to be exactly the same as the drag area.

**Comment by dtarb** [comment]
For exactly overlapping resources, I think that having one box that you can click on to see the 3 resources listed is better than having no box and having to find the resource in a list at the top.  

For zoom control, I understand your explanation.  I just think that the level to which it automatically zooms in should not be so small.

**Comment by xuezhaokun** [comment]
@dtarb Yes. I agree with the suggestion on overlapping problem. I will take a look at the zoom control issue to see if there is any proper modification I can make on it.

**Comment by xuezhaokun** [comment]
@dtarb @danames I am not sure whether we still want to keep working on the new design. Based on my implementation so far, it looks really messy. This is the view when we initialize the map:
![image](https://cloud.githubusercontent.com/assets/6678333/17865831/0fa33334-6872-11e6-8f43-30d4b31c5041.png)
If we zoom to the United States, it looks awful. I think the overlapping boxes just make the map-view even more confused.
![image](https://cloud.githubusercontent.com/assets/6678333/17865859/3c906b1e-6872-11e6-88c4-5a3595a46a11.png)

As the number of box resources increases, it will look worse. I do not think people would like this idea. I am wondering if there is any improvement we could do based on our last design. I think that one looks better than this.

**Comment by dtarb** [comment]
I think that a softer color, like light blue, light green or light orange would be better than the glaring red.  Dont make it darker and darker as you get more overlaps.

Also I think that one of Dan's suggestions was to only show boxes that fit completely within the view.  This could be extended to boxes that have their center within the view.  

And I would consider showing clustered point resources in the same color as point coverages, and box centers and clustered box center resources within the view as the same color, but different from point coverages.

**Comment by xuezhaokun** [comment]
I am not sure whether we could find a way to avoid the boxes get darker and darker as they overlap. I will try to see what I can do here. For now, we are not showing the centers of boxes. I am not quite sure whether people want to get the centers back.

**Comment by xuezhaokun** [comment]
@dtarb After some discussion with Martin, we still think it is better to keep the list-view and map-view tabs. Instead of combining them together, we add a similar grid under the map, which shows all box resources overlap current map bound. If we combine these two views in one tab, users always have to do a filter action when they want to display the mappable results on the list view. We now always provide this filtered list under the map. Users can select mappable resource from this list. I have pushed the latest code to my PR #1388 . I do not how to deploy it on playground. Can you let someone to deploy it on playground and do some test on it? If people like this idea, I will continue the optimization work on this version.

**Comment by dtarb** [comment]
OK.  @mjstealey can deploy for you or tell you the instructions for you to deploy yourself on one of the test VMs.

**Comment by mjstealey** [comment]
@xuezhaokun, @dtarb - deployed to [https://playground.hydroshare.org](https://playground.hydroshare.org)

Branch [discovery-add-geocoder-with-new-style](https://github.com/hydroshare/hydroshare/tree/discovery-add-geocoder-with-new-style) using copy of production resources

**Comment by alvacouch** [comment]
@xuezhaokun @danames @horsburgh @dtarb 

Keep in mind that we are trying to solve an open research problem. There is a new version that does some of what you suggest. 

I would suggest that we stop when this is "good enough" and go on to other pressing issues: 
a) discovery based upon file metadata. 
b) user discovery
c) group discovery
We can always come back to this later. Methinks there is no easy solution that actually "solves" the current problem, because
a) most current solutions assume a bounded collection that is targeted at a single experiment/problem. 
b) we actually have an unbounded collection that lacks that kind of cohesion.

**Comment by xuezhaokun** [comment]
@mjstealey I fixed some bugs on this branch. Can you help me to deploy the latest version of this branch to playground? @dtarb Can you just let people to test it and leave comments on it?

**Comment by mjstealey** [comment]
> deploy the latest version of this branch to playground?

@xuezhaokun - This has been redeployed to playground.

**Comment by xuezhaokun** [comment]
@mjstealey Thanks!

**Comment by dtarb** [comment]
Looking at on Playground the first thing that puzzled me was I see a map with 18 resources centered over Utah, 11 over NC both orange and two over Germany, red dot, but the list below lists 94.  18+11+2 != 94.

There is no indication of this on the page, but I figure out by clicking on items on the list below that they represent box coverages.  I can click on list items one at a time to get the box displayed, but scrolling the list and clicking one at a time is very unwieldy.

So, while I am very sympathetic to moving on, this does not resonate with me. 

I actually prefer the current deployment on www that shows the centers of all box coverages on the map and then the box when you click on the center.  I suggest you actually go back to what is on www and work to enhance it from there.  

Here are enhancements I would suggest starting from www.  Once a resource is selected by selecting the box center there should be an option to zoom to extent to see its coverage.  There should also be the ability to select resources actively by clicking a check box in the pop up or selecting with a select box, rather than automatically going with the map display.  The reason for this is so that when you zoom out to a big box you do not lose the resources you may already have selected.  I would also prefer to see the list of map selected resources below the map, rather than in the list tab, so that effectively the toggle Map results and all results is the tab switch.

**Comment by xuezhaokun** [comment]
@dtarb OK. I will discuss about it with Alva and see what we can do from that point.

**Comment by alvacouch** [comment]
I really don't like just having markers for the extents. The marker can be off the page, so that one has no indication that there is an extent. One option I have discussed with @xuezhaokun is to draw the borders of the extents without shading. This indicates that something is there even if the marker for the extent is off the screen.

**Comment by xuezhaokun** [comment]
@dtarb  I will try to see whether there is some easy way to implement @alvacouch idea on google map. For this idea, I am still not quite sure whether we will have the same performance issue as we had on showing all rectangles on the map. It seems one of the main issues here is whether we want to show box coverages' centers on the map. If not, how can we indicate users that we have box coverage resources and they can discover them on map... If we list all box resources as we have now, can we just make some help text to indicate users what the grid means and what they can do from the grid?

**Comment by dtarb** [comment]
I think that having to click on a list to show a coverage is a bit clumsy.  My first preference would be to try what I suggested above with box centers in a different color, like it is on www right now and then only see the box when a user clicks on the center point.  @alvacouch does not like this due to box centers out of the extent shown not being able to be discovered.  One way to overcome this may be to add a link/button "Expand to show resources whose coverage overlaps the domain".  This button would expand the map to all resources meeting the filter criteria that overlap the domain, and show their centers.  Their coverage could then be shown by the same clicking option.  The search extent would remain the same, i.e. be smaller than the view extent (and be indicated by say a dashed line) and other resources in the expanded view extent that do not touch the search domain would not be shown.

**Comment by alvacouch** [comment]
@dtarb Your suggestion to "expand to show resources whose coverage overlaps the domain" does not make logical sense. Since the domain is the displayed region, this will change the domain. Thus, recursively, you'll end up with the whole world. 

There is something more elegant that takes a bit more time to do. I have seen GIS apps that deal with overlapping domains by shading either the inside or outside of the borders of overlapping regions without shading the interior. Thus, the outlines are shown, but the inside of an outline is obvious from context.  For example, using a gradient rectangle, one gets: 

![gradient](https://cloud.githubusercontent.com/assets/3342292/18216194/58fd901a-7123-11e6-8a80-982cb605ce33.png)

**Comment by dtarb** [comment]
@alvacouch you only read half of my suggestion.  "The search extent would remain the same, i.e. be smaller than the view extent (and be indicated by a dashed line)" separates search and view extent and stops the recursion.  Though if you do have one resource that covers the globe the first expand would go to the globe.  This may not be that bad as it would reflect clearly discovery of global resources and zooming back in will find other resources.  The suggestion does require separation of search extent and view extent though.

I think that shaded rectangles may work.  I look forward to the demo.

**Comment by alvacouch** [comment]
@dtarb @danames @xuezhaokun @horsburgh When Zhaokun and I were talking after the Wed meeting, I had a different and very simple idea for dealing with scale on the coverages. 

1. First, make it possible for a user to drop a pushpin anywhere on the map. Then make the bottom view consist of all coverages overlapping that pushpin. This includes coverages that are larger than the current view. 
2. Depict all coverages as mostly transparent outlines. Thus, if there is a lot of data for a coverage, the outline will be darker.  
3. Hovering over a coverage in the bottom view fills its outline. 

What do people think of that idea? It would most definitely deal with scale, which the popup solution doesn't. In particular, there are regions in which one can click (e.g., near the CZOs) in which there will eventually be hundreds of available coverages.

**Comment by alvacouch** [comment]
@dtarb @danames @xuezhaokun @horsburgh I might add that I've considered other options that seem to have incredible downsides. The following don't seem to scale:  

- Shading the interior of rectangles before clicking. 
- Getting a popup in response to an area click.

The proposal is relatively easy to code, because we don't register the individual rectangles for events (another scaling issues) and instead register the canvas as a whole.

**Comment by alvacouch** [comment]
@dtarb @danames @xuezhaokun @horsburgh I think it's really important to remember that in a discovery situation, we are doing to deal with depiction scale issues that we will never deal with in visualizing an individual resource. A single point can contain thousands of related resources. This is especially true if we eventually give the user the option for searching for "any resource that contains information on this point". This would include model instances as well as point coverages.

**Comment by alvacouch** [comment]
@dtarb @danames @xuezhaokun @horsburgh Of course, we could instead allow the user to drop a rectangle upon the area and report upon that rectangle, but that seems somewhat redundant, as this does the same thing as modifying pan and zoom.  

Your thoughts?

**Comment by alvacouch** [comment]
@xuezhaokun @dtarb @danames @Horsburgh: 

I had a brief conversation with Rick Hooper on this issue. He said that we should not be thinking in GIS terms. Instead, we should be thinking about the discovery problem and how to solve it best. He likes the list of things that show up on the map -- the same ones that have come into contention lately as too "awkward". 

He has the following possible suggestions: 

1.  Perhaps separate the list of extents contained in the view from the list of extents that are larger than the view. 
2. Rectangles are shown for the latter. 
3. Hovering over an element in the list shades its box/polygon. 
(we already know that shading all rectangles under the mouse leads to unusable shadings)

He also suggests that we might be thinking too literally in GIS terms. One alternative to traditional GIS is a "heat map" approach in which we separate the view into squares and shade each square in proportion to the amount of data for the square. Clicking on a square shows its data. Etc.

**Comment by dtarb** [comment]
I think we have circled on this a lot and it's time for you to choose.  Each time I am asked, I find myself coming back to the suggestion already given which are:

- Separation of search extent (a rectangle) from view.  This can be the same as dropping on a search rectangle that Alva suggested, and allows a button "expand to show selected resources" to show the resources found without adding more resources to the found set.
- Showing box centers (as www does now).  This provides a place on the map to click to see details (and the box) for a box coverage resource.
- Show a list of the resources within or overlapping the search extent box.  Clicking on a list element also shows the box.  This fills the list need expressed above and allows identifying resources whose centers may be outside the view extent.  
- Only showing the box for a resource that gets a click (either on its center on the map or on the list).
- These operations are superimposed on the set selected by other non GIS filter facets (author etc. as I think is the case now) so we combine GIS and non GIS discovery.

This has formed itself in my mind as a model that would work.  I am open to other ideas, but each time I see an implementation that does not resonate, this is what I end up coming back to, as a suggestion.

The resource heat map idea does have some appeal, but as some level I think it is helpful to see the resource box and have a place to click (its center) to go to the landing page.  The heat map could be a mode that the map could be placed in (and if we monitor and record which mode people tend to use we would learn which is working better for users).  Could monitoring and learning from user behavior be a functionality that we add for HydroShare 2.0

**Comment by alvacouch** [comment]
@danames @mseul @dtarb We now have a super-elegant solution in which 
* Click always selects what's under the click.
* Drag always pans. 
* Rectangles are selected with click, like points. 

We are seriously considering whether dragging rectangles is needed, because 
* This is a different definition of what drag does. 
* That, in turn, requires a mode button to enable both pan and rectangle select. 
(That mode button would have a hand and an arrow (pan versus select)). 

Argument for the mode button: We can do rectangle selection. 
Argument against the mode button: Simpler user interface. 
(Obviously, if we do the mode button, select is going to be the default) 

Your thoughts? 

In other words, how important is it to be able to select a group of resources to show in the bottom box, and what kinds of things would we want to do to them as a group? 

- @alvacouch and @xuezhaokun

**Comment by alvacouch** [comment]
Oops. Pan should be the default mode. This means that click always means select, but drag won't mean select unless we disable pan.

**Comment by dtarb** [comment]
I do not follow exactly what you are suggesting so perhaps should wait to see a demo.  I do not understand "Pan should be the default" and "click always means select" because to me pan means clicking at a point and then dragging the map.  How does one pan without clicking if click always means select.

I also think it is relatively natural to, when presented with a map with multiple dots, to be able to select more than one of them and do something with all the items selected.  In this case the things to do may be to add to a folder in my resources, or tag with a tag in my resources as ways to categorize and organize the things found.  If these things have to be done one resource at a time, it is OK, but not quite as slick.

**Comment by pkdash** [comment]
@xuezhaokun Can this be closed?

**Comment by alvacouch** [comment]
Apparently we can't close it because there is still content under discussion.

**Comment by dtarb** [comment]
I agree with this being closed as the noise to signal ratio in here is high.  There are many comments about test deploys that no longer exist, so the discussion needs to move on from this issue.  Lets work on map improvements in #1755 and other Discovery improvements in separate issues for them.
